File: PracticaExamen/AccesoDatos/DAOProducto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ObjetosTransferencia;
using System.Data.SqlClient;
using System.Data;

namespace AccesoDatos
{
    public class DAOProducto
    {

        private ConnectionManager connection;
        private SqlCommand comando;

        public DAOProducto() {
            connection = new ConnectionManager();
        }

        public TOProducto search(String cod) {

            TOProducto p = new TOProducto();
            comando = this.searchStatement(cod);
            p.Codigo = cod;
           
            try
            {
                connection.abrir();
                SqlDataReader reader = comando.ExecuteReader();

                if (reader.HasRows)
                {
                    while (reader.Read())
                    {
                        String descripcion;
                        String precio;

                        descripcion = reader["Descripcion"].ToString();
                        precio = reader["PrecioUnitario"].ToString();
                        p.Descripcion = descripcion;
                        
                        p.Precio = int.Parse(precio, System.Globalization.NumberStyles.Currency);
                    }

                }
                
                reader.Close();
                return p;

            }
            catch (SqlException ex)
            {

                throw new Exception(ex.Message);
            }
            finally {
                connection.cerrar();
                
            }

        }

        private SqlCommand searchStatement(String p) {
            String query = "Select * From Productos where Codigo=@cod";

            SqlCommand c = new SqlCommand(query, connection.connection);
            c.Parameters.AddWithValue("cod", p);
            return c;
        }

    }
}

File: Practicav2/AccesoDatos/DAOProducto.vb
Imports ObjetosTransferencia
Imports System.Data
Imports System.Data.SqlClient

Public Class DAOProducto
    Dim conex As ConnectionManager
    Dim comando As SqlCommand

    Sub New()
        conex = New ConnectionManager
    End Sub

    Function buscar(cod As String) As TOProducto
        comando = Me.buscarCommand(cod)
        Dim codigo, descripcion As String
        Dim precio As Integer

        Try
            conex.abrir()
            Dim reader As SqlDataReader = comando.ExecuteReader()

            If reader.HasRows Then
                While reader.Read
                    codigo = reader("Codigo").ToString()
                    descripcion = reader("Descripcion").ToString()
                    precio = Int32.Parse(reader("PrecioUnitario").ToString(), Globalization.NumberStyles.Currency)
                    Exit While
                End While
            End If
            reader.Close()
            Return New TOProducto(codigo, descripcion, precio)
        Catch ex As SqlException
            Throw New Exception(ex.Message)
        Finally
            conex.cerrar()
        End Try

    End Function

    Function guardar(p As TOProducto, num As Integer) As Boolean

        conex.abrir()
        comando = Me.insertCommand(p, num)

        Dim i As Integer = comando.ExecuteNonQuery()

        conex.cerrar()

        Return i <> 0


    End Function

    Private Function buscarCommand(cod As String) As SqlCommand
        Dim query As String = "SELECT * FROM Productos WHERE Codigo=@cod"

        Dim c As SqlCommand = New SqlCommand(query, conex.connection)
        c.Parameters.AddWithValue("@cod", cod)

        Return c
    End Function

    Private Function insertCommand(p As TOProducto, num As Integer) As SqlCommand
        Dim query As String = "Insert into DetalleFactura Values(@num, @cod, @des, @precio)"
        Dim c As SqlCommand = New SqlCommand(query, conex.connection)

        c.Parameters.AddWithValue("num", num)
        c.Parameters.AddWithValue("cod", p.Codigo)
        c.Parameters.AddWithValue("des", p.Descripcion)
        c.Parameters.AddWithValue("precio", p.Precio)

        Return c
    End Function
End Class
